### ISSUE TITLE
Add a tutorial for the fullspectrum rendering

### ISSUE THREAD

**Issue by Flamaros** [GPUOpen-LibrariesAndSDKs/RadeonProRenderSDK#17]
I am trying to make the rasterization backend working without success, it would pretty nice to have a tutorial that explain how set up a viewport with the fullspectrum rendering feature.

**Comment by jrz371** [comment]
Hi,

I've managed to get it working. It's a not much of a hassle but it behaves differently than the full path tracing rendering. It's mainly a large number of unsupported features. Wrapping all your ProRenderSDK calls in the check macro/function will help you find what is and isn't supported.

I use the C# SDK so that's what my snippets will be in. You can probably pretty quickly find the corresponding functions, constants in the C++ SDK.

First, you need to register the Hybrid.dll as a plugin.

`
int[] plugins = { Rpr.RegisterPlugin("Hybrid.dll") };
`

Then you can continue with your context creation normally. After you've create the context you need to set which level you're doing for the hybrid rendering.

`
//Gotten from RadeonProRenderBaikal.cs
public const uint RPR_RENDER_QUALITY_LOW = 0u;
public const uint RPR_RENDER_QUALITY_MEDIUM = 1u;
public const uint RPR_RENDER_QUALITY_HIGH = 2u;
public const uint RPR_RENDER_QUALITY_ULTRA = 3u;

Rpr.Check(Rpr.ContextSetParameterByKey1u(context, (Rpr.ContextInfo)RprBaikal.RPR_CONTEXT_RENDER_QUALITY, (uint)RPR_RENDER_QUALITY_LOW));
`

Hope that helps

**Comment by Flamaros** [comment]
I previously posted on the [forum](https://community.amd.com/t5/sdk-discussions/how-to-activate-rasterization-rendering/m-p/434323/highlight/false#M31)

For me it fails on the creation context:
`rpr_int pluginID = rprRegisterPlugin("Hybrid.dll");
CHECK_NE(pluginID, -1)
rpr_int plugins[] = { pluginID };
size_t pluginCount = sizeof(plugins) / sizeof(plugins[0]);
// Create context using a single GPU 
status = rprCreateContext(RPR_API_VERSION, plugins, pluginCount, g_ContextCreationFlags, NULL, NULL, &g_context);
CHECK(status);
`

I am developing in C++ but the code should be pretty similar.
I suspect that my hardware doesn't match the requirement, but I can't be sure without better error message.

I have tested on a laptop with a gefore 1050 Gtx laptop and a desktop with a geforce 980 GTX (latest version of drivers on both computers). Even if it will not be able to perform at the best level on Nvidia hardware I am expecting that it works (else I can't do a commercial product).

I also check to the RPR_CONTEXT_RENDER_QUALITY parameter that isn't defined in the C++ API, that is pretty curious because it seems that the blender plugin have a render quality that should match this parameter.

I am using the latest version of the RPR SDK, maybe I shouldn't?

Thank you for your help.

**Comment by bsavery** [comment]
> I have tested on a laptop with a gefore 1050 Gtx laptop and a desktop with a geforce 980 GTX (latest version of drivers on both computers). Even if it will not be able to perform at the best level on Nvidia hardware I am expecting that it works (else I can't do a commercial product).

Depending on your hardware you might need to set it to use < 4gb of memory.  You can do this by setting the context properties like we do in the blender plugin: 

https://github.com/GPUOpen-LibrariesAndSDKs/RadeonProRenderBlenderAddon/blob/fb1c0958d289c687c832aa7428fd5a1b41e93dcf/src/rprblender/properties/render.py#L556 Obviously this is in python but you get the idea

**Comment by bsavery** [comment]
`if self.render_quality in ('LOW', 'MEDIUM', 'HIGH') and self.hybrid_low_mem:
            # set these props to use < 4gb
            vertex_mem_size = pyrpr.ffi.new('int*', 768 * 1024 * 1024)  # 768mb texture memory
            acc_mem_size = pyrpr.ffi.new('int*', 1024 ** 3)             # 1gb for bvh memry
            context_props.extend([
                pyrpr.CONTEXT_CREATEPROP_HYBRID_VERTEX_MEMORY_SIZE, vertex_mem_size,
                pyrpr.CONTEXT_CREATEPROP_HYBRID_ACC_MEMORY_SIZE, acc_mem_size])`

**Comment by jrz371** [comment]
@Flamaros I'm using the latest. The file you need to include is RadeonProRender/inc/RadeonProRender_Baikal.h. That has all those macros you need.

**Comment by Flamaros** [comment]
It seems interesting, but I should set those properties only after the context creation, isn't it?

**Comment by jrz371** [comment]
No, you can pass those in during context creation with the props parameter

```
extern RPR_API_ENTRY rpr_status rprCreateContext(rpr_int api_version, rpr_int const * pluginIDs, size_t pluginCount, rpr_creation_flags creation_flags, rpr_context_properties const * props, rpr_char const * cache_path, rpr_context * out_context);
```

**Comment by Flamaros** [comment]
oh, I hope it will help me to make it work.
Thank you both of you for this precious help, I'll try this later.

**Comment by Flamaros** [comment]
I finally got the context creation work :-)

I have an error on CHECK(rprMaterialSystemCreateNode(g_matsys, RPR_MATERIAL_NODE_DIFFUSE, &g_diffuse));
but I should be simpler to fix.

Thank you both again.

Next step will be to figure out how to switch between Hybrid.dll and Northstar64.dll backends.

**Comment by jrz371** [comment]
Only the uber shader is supported for Hybrid rendering. You'll need to convert your existing materials to it in some way.

**Comment by bsavery** [comment]
Correct.  The simple way in this case would be to use the uber shader and only set the Diffuse weight while turning off Reflection weight.

**Comment by Flamaros** [comment]
Ok, nice. It will not be an issue for me at the moment I can render physically based materials and some diffuse for the GUI.

I already have plugged dear Imgui on top of a demo with the OpenGL interop (works fine with path tracer backend).

Now I will be able to tests other mandatory features I need for my project. Normally RPR fit all my needs from what I saw on presentation resources.

Is there an in depth documentation that explains more things than the API reference?
Like limitations of backends and specific API, backend internals, graphics API interops, ...

Can the integrated GPU be selected? GPU0 on a laptop isn't the integrated one on an hybride system?

**Comment by jrz371** [comment]
There are the docs here
https://radeon-pro.github.io/RadeonProRenderDocs/en/sdk/about.html

I'm not sure on integrated devices. We iterate over all the device flags and see what the system has then present that to the user so they can decide which hardware to use.

**Comment by Flamaros** [comment]
I'll do some tests with AmdPowerXpressRequestHighPerformance  and Nvidia equivalent, as I am targetting casual user I don't really want to have complicated settings exposed throw my application UI.

At least it will be nice to add comments about the 4Gb VRAM restriction for the hybrid backend in the documentation and how to change that with the context creation properties.

Maybe we can do pull requests to help to improve the documentation?

**Comment by bsavery** [comment]
Thanks!  I will talk to the docs team about getting this in.  But for now I'm going to close the issue.
